### ISSUE TITLE
A fairly complete explanation about DetectNet input and output layers

### ISSUE THREAD

**Issue by szm2015** [NVIDIA/DIGITS#1011]
Hi everyone
I'm trying to train DetectNet with my own custom data and so am looking for a guide to rightly customize the network for my purpose, I have images of 320 by 240 and the object width I wish to detect vary from 20 to about 80 pixels. The annotation is like this example:

Car 0.0 0 0.0 221.00 54.00 260.00 91.00 0.0 0.0 0.0 0.0 0.0 0.0 0.0 
Car 0.0 0 0.0 207.00 121.00 264.00 175.00 0.0 0.0 0.0 0.0 0.0 0.0 0.0 
Car 0.0 0 0.0 147.00 106.00 200.00 154.00 0.0 0.0 0.0 0.0 0.0 0.0 0.0 
Car 0.0 0 0.0 1.00 148.00 75.00 208.00 0.0 0.0 0.0 0.0 0.0 0.0 0.0 

I've checked [DetectNet: Deep Neural Network for Object Detection in DIGITS](https://devblogs.nvidia.com/parallelforall/detectnet-deep-neural-network-object-detection-digits/) and from what I have found "DIGITS overlays the image with a regular grid with spacing slightly smaller than the smallest object we wish to detect" (this smallest object I have set to 20 in "Minimum box size (in pixels) for validation set" when creating LMDBs). 

Now for training I followed [Deep Learning for Object Detection with DIGITS](https://devblogs.nvidia.com/parallelforall/deep-learning-object-detection-digits/) except the fact that I didn't resize my data to 1536 by 1024 because it couldn't fit in my GPU's memory (4GB) even with batch size set to 1. Here it's mentioned: "Note that in order for DetectNet to successfully detect objects in an image, the object size should be between 50×50 and 400×400 px in the input images." I wonder where this number comes from?! and if that's the case why we set  "Minimum box size (in pixels) for validation set" and also the stride which is set to 16 by default.

So, I just copied detectnet_network.prototxt as the network definition but I changed the default x and y sizes (1248 and 384) to my own data size, meaning I changed these line:
79: image_size_x: 1248
80: image_size_y: 384

118: image_size_x: 1248
119: image_size_y: 384

2504: param_str : '1248, 352, 16, 0.6, 3, 0.02, 22, 1'
2519: param_str : '1248, 352, 16, 1'
2545: param_str : '1248, 352, 16'

But to be honest I don't understand the last 3 ones (and Why 352 is used instead of 384?) the other parameters are set as mentioned in [Deep Learning for Object Detection with DIGITS](https://devblogs.nvidia.com/parallelforall/deep-learning-object-detection-digits/) and up to now (epoch 300 with batch size 10 and batch accumulation 2) I'm getting very low recall and mAP (1.6). Here is my loss graph (I omitted the parts with very large train loss so that mAP is recognizable)
![loss](https://cloud.githubusercontent.com/assets/16157023/18033891/1fa03422-6d44-11e6-85f9-eff598075a9b.jpg)
This the entire loss graph:
![loss_whole](https://cloud.githubusercontent.com/assets/16157023/18033947/41736ba4-6d45-11e6-8c3a-ec7cbb2a82da.jpg)

I tried to test an image by this network I trained and I got no bounding boxes, however the heat map below (I believe that's what it's called) fits the objects very well,
![heatmap](https://cloud.githubusercontent.com/assets/16157023/18034021/65cd0594-6d47-11e6-9115-91ecec1e7bfc.jpg)

Sorry that it got too long, but this is the end! Now to summarize my questions:
    1. What is the "detector's working area" (the 50 by 50 and 400 by 400 numbers) and how to calculate 
        and even change it?!
    2. what is 'param_str' in lines 2504, 2519, 2545 of detectnet_network.prototxt, which I mentioned 
        above? and how to change them to fit our data? (this question have also been asked in  #980)
    3. Other related points in order to better customize detectNet for other data sets?!
    4 and Last. why does the heat map shows the objects but the are no bounding boxes listed?

Thank you so much for your help.

**Comment by szm2015** [comment]
Ok, I tried the thing suggested by jbarker-nvidia in #980 (setting crop_bboxes: false) and it improved the mAP from 1.6 to 14 percent after epoch 400 (though far from desirable accuracy it still shows significant improvement), also I was able to get bounding boxes at inference time. but I still fail to see why the heat map I attached before does not result in detections.

Also as the topic suggests as my main question, I want to find out more about the network so I can further improve my results, and so am I correct in assuming that by setting crop_bboxes to false we invoke random cropping?

My other questions still hold!

**Comment by jbarker-nvidia** [comment]
@szm2015 some answers to your questions:

1.  The network has a receptive field of 555x555 pixels with a stride of 16 pixels. This means that each cell in the output grid is influenced by 555x555 pixels in the input image.  The 50 and 400 numbers for object size are approximate and come from experience with DetectNet rather than a precise calculation.  For your problem where the object sizes are smaller I would recommend that you increase your training image sizes (perhaps double them) and then use training time cropping to the original image size (320x240) to keep memory size down.  This will bring your target objects in to the range that DetectNet is sensitive to.

2. The 'param_str' parameters are lists of parameters used by the `clustering.py`and `mean_ap.py` functions found in `$NVCAFFE_HOME/python/caffe/layers/detectnet`.  The comments in those Python scripts explain the parameters - in short, they are the output grid size and stride and parameters used by the OpenCV function GroupRectangles that is used to cluster the output bounding boxes produced by DetectNet.

3.  If you are not currently using exponential learning rate decay I recommend trying that with a decay rate of 0.99.  Other than that the required customizations are dataset and target object size dependent.  It is possible to reduce the stride of DetectNet, which may help for your problem, but it will also greatly increase GPU memory usage so may not work for you.

4. The heatmaps you show are actually the predicted offset for the bounding box corner at each pixel.  The coverage heatmap above that in the DIGITS output will show the likelihood that an object is present at each location.  The clustering layer that post-processes the DetectNet output applies a threshold based on the coverage value, so even though the heatmaps show that the network is sensitive to the presence of the objects the coverage value may not be exceeding this threshold.

**Comment by szm2015** [comment]
@jbarker-nvidia Hi, thank you for your complete reply, it really enlightened things for me! (specially reading clustering.py and mean_ap.py) 

 I did what you said and increased my image size to 2.5 times the previous ones (now my objects would be between 50 and 200 pixels width). I was also using exponential decay from before, with the same stride of 16 and setting the crop_bboxes: false with image x and y set to 320 and 240), I got better results (recall = 72.5, precision = 32, mAP = 23.7) after 300 epochs and I don't think training it more would do any good as it's been roughly the same number from epoch 200. any idea why the precision is so low?

I must add that I trained for another object on this same size images (object width between 60 and 140 pixels) with the same configuration, this time I got a mAP of 94 percent! well to clarify, the first object I mentioned is a car and this second one a bus. what do you think could be the difference? (Note that I didn't have to resize the images this time as the objects width was well within the range, and so I set crop_bboxes: false with image x and y the same as image width and height)

**Comment by szm2015** [comment]
Where can I find definition of layer "DetectNetTransformation"?

**Comment by jbarker-nvidia** [comment]
[https://github.com/NVIDIA/caffe/blob/caffe-0.15/src/caffe/layers/detectnet_transform_layer.cpp](The code is here)

**Comment by lasandcris** [comment]
@szm2015 - what's the size of your dataset? I also had similar problems to yours and having increased the number of training/val images made a significant difference.

**Comment by szm2015** [comment]
for cars I have 243 images with 755 labeled objects (cars) and for bus I have 482 images with 512 labeled objects (buses). the number of labeled objects is less for bus yet it has a far better result, but none the less I will increase car objects to see whether it makes a difference or not.

**Comment by sherifshehata** [comment]
@jbarker-nvidia  
After checking the .py files, i came to a conclusion that you set the dimension to be 352 not 348 to make it a multiple of the stride. Is this correct?

**Comment by sherifshehata** [comment]
To contribute to the "guide", in my experiments disabling all transformations in the transformation layer helped getting a very good mAP. The used transformations didn't suit my data, so disabling them helped much.

Additionally, @jbarker-nvidia suggestion of using randomly cropped patches helped much in getting better mAP and faster convergence.

**Comment by varunvv** [comment]
**@jbarker-nvidia**
" For your problem where the object sizes are smaller I would recommend that you increase your training image sizes (perhaps double them) and then use training time cropping to the original image size (320x240) to keep memory size down "

Do I have to change the x and y coordinate values in the label file while resizing the image. 
Or just use the resize option in the digits interface?

**Comment by jbarker-nvidia** [comment]
@varunvv DIGITS will automatically resize the labels too. You need the extent of the coordinates in the label files to match the original images that DIGITS ingests at dataset creation time.

**Comment by chentsovv** [comment]
I had a similar problem, I solve by increasing the number of epochs. My net 600 pictures 1024x720, Box~140-170 px. I got mAP after 40 epoch.
